backend/layers/business/business.py

```diff
@@ -1099,6 +1090,57 @@ def delete_artifacts(self, artifacts: List[DatasetArtifact]) -> None:
             except S3DeleteException as e:
                 raise CollectionDeleteException("Attempt to delete public Datasets failed") from e
 
+    def _update_revision_publisher_metadata(self, version: CollectionVersion, date_of_last_publish: datetime) -> None:
+        """
+        Call Crossref for the latest publisher metadata and:
+        - if a DOI has moved from pre-print to published, trigger update to collection (and artifacts) and raise exception,
+        forcing curators to re-publish the collection once artifacts update is complete, otherwise,
+        - if Crossref has been updated since last publish, update collection version publisher metadata.
+
+        :param collection_version_id: The collection version to check publisher updates for.
+        :param date_of_last_publish: The originally published at or revised at date of collection version.
+        """
+        # Get the DOI from the collection version metadata; exit if no DOI.
+        links = version.metadata.links
+        link_doi = next((link for link in links if link.type == "DOI"), None)
+        if not link_doi:
+            return
+
+        # Fetch the latest publisher metadata from Crossref. Ignore errors, and exit if no metadata is found.
+        publisher_metadata, crossref_doi_curie, deposited_at_timestamp = self._get_publisher_metadata(link_doi.uri, [])
+        if not publisher_metadata or not crossref_doi_curie:
+            return
+
+        # Handle change in publisher metadata from pre-print to published.
+        crossref_doi = f"https://doi.org/{crossref_doi_curie}"
+        if crossref_doi != link_doi.uri:
+
+            # Set the DOI in the collection version metadata links to be the returned DOI and update collection
+            # version (subsequently triggering update of artifacts).
+            updated_links = [Link(link.name, link.type, crossref_doi) if link.type == "DOI" else link for link in links]
+            update = CollectionMetadataUpdate(
+                name=None,
+                description=None,
+                contact_name=None,
+                contact_email=None,
+                links=updated_links,
+                consortia=None,
+            )
+            self.update_collection_version(version.version_id, update, True)
+
+            # Curators will need to re-publish once artifact updates are complete.
+            raise CollectionPublishException(
```

Comment:
nit: since this function is extracted from the publish function now, we probably shouldn't raise a publish-specific exception directly in it. That way it can be cleanly re-used in other contexts. 

What do you think of returning a bool like "has_updated_doi", where True means we throw the exception in publish_collection_version

Reply:
I'll need the new DOI returned from Crossref in the calling function for the error message. The function could return `Optional[Tuple[str, str]]` so `existing_doi, crossref_doi` if there is a change in DOI, otherwise `None`. Thoughts?